src/sa.py

```diff
@@ -1,15 +1,149 @@
 import argparse
+import fasta, fastq
+import re
 
+def getSuffixes(x: memoryview):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    return suffixes
+
+def getTrailingNumber(s):
+    m = re.search(r'\d+$', s)
+    return (s[:m.start()], int(s[m.start():]))
+
+def strcmp(s1, index, x):
+    max_i = min(len(s1), len(x)-index)
+    i = 0
+    while i < max_i and s1[i] == x[index+i]:
+        i += 1
+        if i >= max_i:
+            return 0
+    if s1[i] > x[index+i]: return 1
+    else: return -1
+
+
+def search(sa, pattern, genome):
+    low = 0
+    high = len(sa)-1
+
+    if low == high:
+        cmp = strcmp(pattern, sa[low], genome)
+        if cmp == 0:
+            yield sa[low] + 1
+            return
+
+    lowCompared = False
+    highCompared = False
+    while True:
+        mid = (high+low)//2
+        midSuffix = sa[mid]
+        cmp = strcmp(pattern, midSuffix, genome)
+
+        if cmp==0:
+            while cmp==0 and len(pattern) > len(genome) - midSuffix:
+                mid += 1
+                if mid >= len(sa): return
+                midSuffix = sa[mid]
+                cmp = strcmp(pattern, midSuffix, genome)
+            
+            oldMid = mid
+            # Compare through both sides for all results
+            while cmp==0:
+                yield sa[mid] + 1
+                mid += 1
+                if mid >= len(sa): break
+                midSuffix = sa[mid]
+                cmp = strcmp(pattern, midSuffix, genome)
+            
+            mid = oldMid - 1
+            if mid < 0: return
+            midSuffix = sa[mid]
+            cmp = strcmp(pattern, midSuffix, genome)
+            while cmp==0:
+                yield sa[mid] + 1
+                mid -= 1
+                if mid >= len(sa): return
+                midSuffix = sa[mid]
+                cmp = strcmp(pattern, midSuffix, genome)
+            return
+        else:
+            if high-low <= 1:
+                # Compare the bounds if they haven't
+                if not(highCompared):
+                    cmp = strcmp(pattern, sa[high], genome)
+                    if cmp == 0:
+                        yield sa[high] + 1
+                if not(lowCompared):
+                    cmp = strcmp(pattern, sa[low], genome)
+                    if cmp == 0:
+                        yield sa[low] + 1
+                return
+            if cmp < 0:
+                high = mid
+                highCompared = True
+            else:
+                low = mid
+                lowCompared = True
```

Comment:
You could simplify this a lot (and speed it up, although not asymptotically) if you used the trick of narrowing the interval one character at a time. For this, you would need a function for finding the lower bound in an interval `lo:hi` where you know that all strings off to an offset, `offset`, are the same, so you just need to look at index `offset` of the strings there:

```
      aaaab...
lo -> abaaa...
      abaab... <- lower('b', lo,hi,offset=4)
      abaab...
      abaac...
hi -> abbaa...
      abbca...
```

Here, all the strings in `sa[lo:hi]` starts with `abaa` and if we search for the lower bound of `b` inside the block at index 4, we find the location pointed at to the right. If you have an `upper_bound()` function as well (and you can get it from the lower bound by adding one to the letter), then you can also get the end of the block of `b` inside `sa[lo:hi]`, and now you have the block that starts with `abaab`.

```
      aaaab...
lo -> abaaa...
      abaab... <- lower('b', lo,hi,offset=4)
      abaab...
      abaac... <- upper('b', lo,hi,offset=4)
hi -> abbaa...
      abbca...
```

Notice that my `upper()` points one past the block of `b`. That's because I wasn't raised by wolves.

These two functions are extremely easy to implement. The `lower_bound()` function is just a binary search where you only move `lo` when it is too small, and you don't stop if the midpoint matches what we search for.  The `upper_bound()` function just calls `lower_bound()` with `chr(ord('b') + 1)` (which is just Python for `'b' + 1`, but necessary to turn the letter `'b'` into a number, 98, so we can add one to it, to 99, and then turn that back into a letter, `'c'`.

```python
def lower(a: str, x: str, as: memoryview, lo: int, hi: int, offset: int) -> int:
    """Finds the lower bound of `a` at `offset` in the block defined by `lo:hi`."""
    while lo < hi:  # Search in sa[lo:hi]
        m = (lo + hi) // 2
        if x[sa[m] + offset] < a: # compare at column offset in sa
            lo = m + 1
        else:
            hi = m
    return lo

def upper(a: str, x: str, as: memoryview, lo: int, hi: int, offset: int) -> int:
    """Finds the upper bound of `a` at `offset` in the block defined by `lo:hi`."""
    return lower(chr(ord(a) + 1), x, sa, lo, hi, offset)
```

The `x[sa[m] + offset]` can go out of bounds, but if you want to avoid having to test for it, you can wrap `x` in a class that returns the sentinel if that happens, or, as we have seen at the exercises, you can index modulo `x`'s length (`x[(sa[m] + offset) % len(x)]`) and you will be fine.

With that, you can essentially just do

```python
lo, hi = 0, len(sa)
for offset, a in enumerate(p):
    lo = lower_bound(a, x, sa, lo, hi, offset)
    hi = upper_bound(a, x, sa, lo, hi, offset)
print("matches:", sa[lo:hi])
```

We move the lower bound in the call to `lower_bound()` before we use it in the call to `upper_bound()`, but that will not affect the result there; the `lo` we get from `lower_bound()` is also a lower bound for the result we would get if we had used the original `lo`, so all is fine.

(Ok, I haven't tested this, but something like it should work).